mlir/include/mlir/Interfaces/DataLayoutInterfaces.td

```diff
@@ -194,6 +194,162 @@ def DataLayoutSpecInterface : AttrInterface<"DataLayoutSpecInterface"> {
   }];
 }
 
+def TargetDeviceDescSpecInterface : AttrInterface<"TargetDeviceDescSpecInterface"> {
+  let cppNamespace = "::mlir";
+
+  let description = [{
+    Attribute interface describing a target device description specification.
+
+    A target device description specification is a list of device properties (key)
+    and their values for a specific device. The device is identified using "device_id"
+    (as a key and ui32 value) and "device_type" key which must have a string value.
+    Both "device_id" and "device_type" are mandatory keys. As an example, L1 cache
+    size could be a device property, and its value would be a device specific size.
+
+    A target device description specification is attached to a module as a module level
+    attribute.
```

Comment:
Can you elaborate on this scenario a bit more? Our thought process currently is that the system desc/device desc would contain properties of the underlying hardware (and hence would suffice to be at the module level). Current implementation is not supporting the idea of cost model (or in other words, the heuristics used by compiler passes). Given that which types of properties would be specialized at a function level?

Reply:
A function could be specialized to use AVX and another for SSE for example.

Look also at the Datalayout analysis in-tree, it kind of achieve this.

Reply:
The way I was thinking this is not to replicate the system descriptor per function, but to have a single (whole) descriptor per module and then an annotation on the op (incl. function) with the identifier (ID or search query) into the module's descriptor.

```mlir
module {
  target = #dlti.system_desc { target[0] = { ... }, target[1] { ... } ... }

  ...
  func.func myFunction(...) #target = 0 {
    ...
  }
  ...
  func.func otherFunc (...) {
    linalg.generic ... #target = 1 {
      ...
    }
  }
  ...
}
```

Reply:
@joker-eph This is the last unaddressed comment. Does my answer help with your question? I think we can solve this later, either relaxing the constraint (as you propose), or creating a hook back to the module's attribute (as I propose). I'd like to get this merged, so we can multi-task the following changes that all depend on this one.

Reply:
Since attributes are owned by context anyway, it doesn't cost us much to allow them at all levels. It would also avoid unchecked indirection (what if someone decides to reorder the list in system description, for example?)

Reply:
Ok, I think we need to have a better API for this anyway, but this PR is already too long to get there. We can have as fewer restrictions as we can right now to not make that harder in the near future.